fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/models/ResourceFactoryImpl.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to DuraSpace under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * DuraSpace licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.kernel.impl.models;
+
+import org.fcrepo.kernel.api.exception.InteractionModelViolationException;
+import org.fcrepo.kernel.api.models.Container;
+import org.fcrepo.kernel.api.models.FedoraBinary;
+import org.fcrepo.kernel.api.models.FedoraResource;
+import org.fcrepo.kernel.api.models.FedoraTimeMap;
+import org.fcrepo.kernel.api.models.FedoraWebacAcl;
+import org.fcrepo.kernel.api.models.NonRdfSourceDescription;
+import org.fcrepo.kernel.api.models.ResourceFactory;
+import org.fcrepo.persistence.api.PersistentStorageSession;
+import org.fcrepo.persistence.api.PersistentStorageSessionFactory;
+import org.fcrepo.persistence.api.exceptions.PersistentStorageException;
+
+
+/**
+ * Implementation of ResourceFactory interface.
+ *
+ * @author whikloj
+ * @since 2019-09-23
+ */
+public class ResourceFactoryImpl implements ResourceFactory {
+
+    private final PersistentStorageSession psSession;
+
+    /**
+     * Special constructor for testing.
+     *
+     * @param psFactory Persistent storage session factory.
+     */
+    protected ResourceFactoryImpl(final PersistentStorageSessionFactory psFactory) {
```

Comment:
how are we expecting psSession to make it into here if this is a test method?

My sense is that we need to pick one of two directions:

1.  ResourceFactory should be a singleton and live for the entire life of the application. In this case it would be passed the current fedora transaction per method call, like how FedoraSession was passed into ContainerService before. This approach would be most appropriate if the factory needs to inject common dependencies in to the constructed resources.
2. ResourceFactory would be scoped to a single transaction. Right now it is scoping to a persistence transaction, but it seems more appropriate to me that its API would expect to work against a FedoraTransaction since it is at the kernel level. This case seems more appropriate if the persistence level is able to inject all of the necessary dependencies into the constructed resources.

This current implementation seems to be assuming that the resources are ready to go after `psSession.read`. Are there any non-persistence level details that might need to be added? At the moment I can mainly think of potentially needing access to the PersistentStorageSessionFactory in order to retrieve triples/content. I wonder if there would be any need for ways to look up other resources by identifier, or pull from indexes?

Reply:
I am perhaps not clear on what we are expecting the PersistentStorageSession to return.

I was thinking _if_ the resource was found then the storage would return the fully formed FedoraResource. If it doesn't return the fully formed object how does another service know where to access the bits to make it fully formed? Because I would assume those "bits" would also be in the persistent storage.

But then if you are `PATCH`ing, you would send this fully formed FedoraResource to some other service to apply new triples or binary or whatever to it.

As for the singleton vs transaction, I am fine with passing in a transaction (`FedoraTransaction` I believe) and generating a readonly session to the persistent storage to retrieve any objects *if* they exist.

Reply:
Also to be clear, each of the calls (ie. `findOrInitContainer()`) should have a `FedoraTransaction` as the first argument?